### PR TITLE
Create GitHub releases in releaser/tag-pusher workflows

### LINKED ISSUE #228: Create GitHub releases in releaser/tag-pusher workflows
The suggestion here is for the [versioning workflows](https://github.com/protocol/.github/blob/master/VERSIONING.md#using-the-versioning-workflows) and [manual version pushes](https://github.com/protocol/.github/blob/master/VERSIONING.md#dealing-with-manual-pushes) to create a GitHub release among with the version tag.

Below, I've gathered some ideas for the process from Slack discussions.

###### How to generate release notes?
- use [auto generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes); [API access to this feature](https://docs.github.com/en/rest/reference/repos#generate-release-notes-content-for-a-release)
- for versioning workflow - we could use the comments/description of the PR updating the version to create the release notes (Q: how to access PR in the workflow?)
- for manual push workflow - we could use the comments/description of the issue that was created as a result of pushing tag manually (Q: should we check if the release was created manually too?)

###### How to verify the GitHub release before subscribers get notified about it?
- initially create the GitHub release in a draft state (make it possible for owners to change this behaviour once they gain enough confidence in the workflow); [API access to this feature](https://docs.github.com/en/rest/reference/repos#create-a-release)

### PR DESCRIPTION
**Pull request by galargh**
Resolved #228

###### Description

This PR builds up on https://github.com/protocol/.github/pull/295. It is currently targeted at that PR. It can be retargeted to `next` once that PR is merged.

In this PR I refactor `release-check` and `releaser` workflows in such a way that enables adding GitHub Release creation to the process. I propose for a draft GitHub Release to be created during `release-check` and for that draft to be published during `release`. `release-check` will post a link to the draft as a comment. The draft can be modified or deleted manually - those actions will be respected by the `release`.

I describe what the refactored workflows are doing below.

###### `release-check`

1. Retrieve version from `version.json` through GitHub API
1. Check if a tag for the retrieved version already exists - exit if it does
1. Check if version is valid
1. Check if a draft GitHub release for the version already exists - create if it does not
1. Find previous version
1. Run diff, `gorelease` and `gocompat` if a previous version exists
1. Post a comment on the PR

###### `release`

1. Retrieve version from `version.json` through GitHub API
1. Check if there exists a PR associated with the commit which is merged and has `release` label on it - exit if not on a default branch and such a PR does not exist
1. Check if a draft GitHub release for the version exists - publish the release if it does, publish the tag otherwise

###### Testing

Default branch

- [x] PR: https://github.com/galorgh/go-ipfs-cmds/pull/5
- [x] release check workflow run: https://github.com/galorgh/go-ipfs-cmds/actions/runs/1894340446
- [x] release workflow run: https://github.com/galorgh/go-ipfs-cmds/actions/runs/1894400874
- [x] release/tag: https://github.com/galorgh/go-ipfs-cmds/releases/tag/v0.7.1

Release branch

- [x] PR: https://github.com/galorgh/go-ipfs-cmds/pull/6

### CODE REVIEW

File: .github/workflows/release-check.yml [hunk 1]
```diff
@@ -2,134 +2,165 @@ name: Release Checker
...
+          # https://docs.github.com/en/rest/reference/repos#get-repository-content
+          version="$(gh api -X GET 'repos/${{ github.event.pull_request.head.repo.full_name }}/contents/version.json' -f ref='${{ github.event.pull_request.head.sha }}' --jq '.content' | base64 -d | jq -r '.version // ""')"
+          status=$? && ([[ $status == 0 ]] || exit $status)
           echo "version=$version"
-          echo "VERSION=$version" >> $GITHUB_ENV
-      - name: Check if the tag already exists
-        # Check if a git tag for the version (as read from version.json) exists
-        # If that is the case, we don't need to run the rest of the workflow.
-        if: env.VERSION != ''
+          echo "::set-output name=this::$version"
```
Comment (laurentsenta) [inline .github/workflows/release-check.yml:20]: Should we use `echo "{name}={value}" >> $GITHUB_OUTPUT` ?

https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

File: .github/workflows/release-check.yml [hunk 2]
```diff
@@ -2,134 +2,165 @@ name: Release Checker
...
-          status=0
-          git rev-list $VERSION &> /dev/null || status=$?
-          if [[ $status != 0 ]]; then
-            echo "TAG_EXISTS=false" >> $GITHUB_ENV
-          fi
-      - name: Install semver (node command line tool)
-        if: env.TAG_EXISTS == 'false'
-        run: npm install -g "https://github.com/npm/node-semver#e79ac3a450e8bb504e78b8159e3efc7089569" # v7.3.5
+          # https://docs.github.com/en/rest/reference/git#get-a-reference
+          gh api '/repos/${{ github.repository }}/git/ref/tags/${{ steps.version.outputs.this }}' && tag_exists='true' || tag_exists='false'
```
Comment (laurentsenta) [inline .github/workflows/release-check.yml:26]: We're using string interpolation at the yml level here, is it possible a malformed value in `steps.version.outputs.this`  turns into a string injection?

Should we use something like:

```yml
env:
    TAG_URL: /repos/${{ github.repository }}/git/ref/tags/${{ steps.version.outputs.this }}
run:
    gh api "${TAG_URL}" # using shell's expansion here
```

?

ref: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

File: .github/workflows/release-check.yml [hunk 3]
```diff
@@ -2,134 +2,165 @@ name: Release Checker
...
+          status=$? && ([[ $status == 0 ]] || exit $status)
+
+          if [[ -z "$release" ]]; then
+            # https://docs.github.com/en/rest/reference/releases#create-a-release
+            # creating a draft release does not create a tag, only publishing does
+            release="$(gh api '/repos/${{ github.repository }}/releases' -F 'draft=true' -f 'tag_name=${{ needs.prepare.outputs.version }}' -F 'generate_release_notes=true' --jq '.html_url')"
+            status=$? && ([[ $status == 0 ]] || exit $status)
+          fi
+
+          echo "COMMENT<<EOF
```
Comment (laurentsenta) [inline .github/workflows/release-check.yml:58]: TIL... cool pattern :)

Do we need to worry about script injection here?
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings

**Comment by galargh** [comment]
Moved back to draft for now. I'm going to asses the comments from @laurentsenta and mark it as ready for review afterwards. Thanks 🙇

**Comment by galargh** [comment]
Replaced this one with https://github.com/protocol/.github/pull/456

**Pull request closed without merging**